src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -12,99 +13,322 @@ namespace System.Net.Sockets
 {
     internal sealed unsafe class SocketAsyncEngine
     {
+        //
+        // Encapsulates a particular SocketAsyncContext object's access to a SocketAsyncEngine.  
+        //
+        public struct Token
+        {
+            private readonly SocketAsyncEngine _engine;
+            private readonly IntPtr _handle;
+
+            public Token(SocketAsyncContext context)
+            {
+                AllocateToken(context, out _engine, out _handle);
+            }
+
+            public bool IsAllocated
+            {
+                get { return _engine != null; }
```

Comment:
Wanted to avoid mutating the state in Token.  Maybe this should be `WasAllocated`? :)